common/check_obsolete_replicas

```diff
@@ -1,84 +1,53 @@
-                          SELECT
-                              id AS rse_id
-                          FROM
-                              atlas_rucio.rses
-                          WHERE
-                              deleted=0) a
-                  LEFT OUTER JOIN
-                      (
-                          SELECT
-                              /*+  INDEX_FFS(replicas REPLICAS_TOMBSTONE_IDX) */
-                              rse_id,
-                              COUNT(1)   AS files,
-                              SUM(bytes) AS bytes
-                          FROM
-                              atlas_rucio.replicas
-                          WHERE
-                              (
-                                  CASE
-                                      WHEN tombstone IS NOT NULL
-                                      THEN rse_id
-                                  END) IS NOT NULL
-                          AND tombstone=to_date('1-1-1970 00:00:00','MM-DD-YYYY HH24:Mi:SS')
-                          GROUP BY
-                              rse_id) b
-                  ON
-                      a.rse_id=b.rse_id)
-
-        LOOP
-                MERGE INTO atlas_rucio.RSE_USAGE
-                USING DUAL
-                ON (atlas_rucio.RSE_USAGE.rse_id = u.rse_id and source = 'obsolete')
-                WHEN NOT MATCHED THEN INSERT(rse_id, source, used, files, updated_at, created_at)
-                VALUES (u.rse_id, 'obsolete', u.bytes, u.files, u.updated_at, u.updated_at)
-                WHEN MATCHED THEN UPDATE SET used=u.bytes, files=u.files, updated_at=u.updated_at;
-
-                MERGE INTO ATLAS_RUCIO.RSE_USAGE_HISTORY H
-                USING DUAL
-                ON (h.rse_id = u.rse_id and h.source = 'obsolete' and h.updated_at = u.updated_at)
-                WHEN NOT MATCHED THEN INSERT(rse_id, source, used, files, updated_at, created_at)
-                VALUES (u.rse_id, 'obsolete', u.bytes, u.files, u.updated_at, u.updated_at);
-
-                COMMIT;
-        END LOOP;
-END;
-'''
-        SESSION.execute(QUERY)
-    except Exception as error:
-        print error
+        session = get_session()
+        for rse in list_rses(session=session):
+            query = select(
+                func.coalesce(func.sum(models.RSEFileAssociation.bytes), 0).label('bytes'),
+                func.count().label('files')
+            ).with_hint(
+                models.RSEFileAssociation, 'INDEX(replicas REPLICAS_RSE_ID_TOMBSTONE_IDX)', 'oracle'
+            ).where(
+                models.RSEFileAssociation.rse_id == rse['id'],
+                models.RSEFileAssociation.tombstone == OBSOLETE
+            )
+
+            bytes_, files = session.execute(query).one()
+            set_rse_usage(rse=rse['rse'], source='obsolete', used=bytes_,
+                          free=None, issuer='root', files=files,
+                          session=session)
+    except Exception:
+        print(traceback.format_exc())
```

Comment:
should there be a `session.rollback()` here?

Reply:
More importantly, I’m missing a `session.commit()`, so many thanks for your comment.

But I’m not yet convinced what I’m doing here is a good idea. If we didn’t have a custom query, we wouldn’t try to tie the gateway calls to a particular session. Is there a benefit to ensuring either all or none of the RSE usage counters are updated? I’m not sure. My initial motivation was to be able to add a dry-run mode easily, but this may not be the best way to proceed.